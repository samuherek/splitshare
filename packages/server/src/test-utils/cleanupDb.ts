import { getConnection } from 'typeorm';

export default async function cleanupDb() {
  await getConnection().query('DELETE FROM bill_user');
  await getConnection().query('DELETE FROM receipt');
  await getConnection().query('DELETE FROM bill_invite');
  await getConnection().query('DELETE FROM bill');
  await getConnection().query('DELETE FROM "user"');
}
