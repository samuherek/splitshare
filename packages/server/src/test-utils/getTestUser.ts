import { Connection } from 'typeorm';
import { User } from '../features/user/entity';

export const testUser = {
  email: 'user@email.com',
};

export async function getTestUser(conn: Connection) {
  const user = await conn.getRepository(User).findOne(testUser);

  if (!user) {
    throw new Error('No Test user found');
  }

  return user;
}
