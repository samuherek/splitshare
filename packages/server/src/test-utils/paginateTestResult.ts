import { PaginationConnection } from '../types';

export default function paginateTestResult(
  resArray: any[]
): PaginationConnection<any> {
  return {
    edges: resArray.map((r) => ({ node: r, cursor: 'nodeCursor' })),
    pageInfo: {
      hasNextPage: false,
      endCursor: 'endCursor',
      itemsCount: 0,
    },
  };
}
