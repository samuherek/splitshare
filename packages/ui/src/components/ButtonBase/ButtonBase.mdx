---
name: ButtonBase
---

import { Playground, PropsTable } from 'docz';
import ButtonBase from './ButtonBase';

# Button

### Props

<PropsTable of={ButtonBase} />

### Playground

<Playground>
  <ButtonBase>ButtonBase</ButtonBase>
</Playground>
