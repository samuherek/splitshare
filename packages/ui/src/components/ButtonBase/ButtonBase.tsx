// @flow
import * as React from 'react';
import styled, { css } from '../../styles/styled-components';

export interface IBaseButtonProps {
  children: Node;
  onClick?: any;
  component: React.ComponentType | 'button' | 'a';
  to?: string;
  disabled?: boolean;
  tabIndex?: string | number;
  href?: string;
  type?: string;
  withNewTab?: boolean;
  className?: string;
}

export interface IButtonHTMLProps {
  className?: string;
  disabled?: boolean;
  type?: string;
  role?: string;
}

// TODO: Fix the ts types
const ButtonBaseStyled: any = styled.button`
  color: inherit;
  cursor: pointer;
  margin: 0;
  border: 0;
  display: inline-flex;
  outline: none;
  position: relative;
  user-select: none;
  align-items: center;
  border-radius: 0;
  vertical-align: middle;
  justify-content: center;
  text-decoration: none;
  background-color: transparent;
  font: inherit;
  padding: 0;

  &:disabled {
    cursor: default;
    pointer-events: none;
    box-shadow: none;
  }

  &:active {
    transform: translateY(1px);
  }

  /* This is for html tags which don't support ":disabled" for the time being*/
  ${({ disabled }) => {
    if (disabled) {
      return css`
        cursor: default;
        pointer-events: none;
        box-shadow: none;
      `;
    } else {
      return null;
    }
  }} /* &:focus {
    outline: 0;
  } */
`;

class Button extends React.PureComponent<IBaseButtonProps, void> {
  static defaultProps = {
    tabIndex: '0',
    component: 'button',
    to: '',
  };

  handleClick = (ev: React.SyntheticEvent<HTMLButtonElement>) => {
    const { onClick } = this.props;
    ev.currentTarget.blur();

    if (typeof onClick === 'function') {
      onClick(ev);
    }
  };

  render() {
    const {
      children,
      component,
      tabIndex,
      className,
      disabled,
      to,
      ...other
    } = this.props;

    let ComponentProp = component;

    const buttonProps: IButtonHTMLProps = {};
    buttonProps.className = className;
    buttonProps.disabled = disabled;

    // Sometimes it's helpful to use dataset props. Thus we need to make sure we apply them to the button
    Object.keys(other).forEach(key => {
      if (key.startsWith('data-')) {
        buttonProps[key] = other[key];
      }
    });

    if (ComponentProp === 'button' && other.href) {
      ComponentProp = 'a';
    }

    if (ComponentProp === 'button') {
      buttonProps.type = other.type || 'button';
      buttonProps.disabled = disabled;
    } else {
      buttonProps.role = 'button';
    }

    return (
      <ButtonBaseStyled
        onClick={this.handleClick}
        as={component}
        tabIndex={disabled ? -1 : Number(tabIndex)}
        to={other.href || to}
        href={component === 'a' && other.href ? other.href : null}
        target={other && other.withNewTab ? '_blank' : null}
        {...buttonProps}
      >
        {children}
      </ButtonBaseStyled>
    );
  }
}

export default Button;
