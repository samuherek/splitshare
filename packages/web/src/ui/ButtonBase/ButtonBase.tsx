import { Link } from "@reach/router";
import clsx from "clsx";
import React, { SyntheticEvent } from "react";
import styled from "styled-components";

export type BaseButtonProps = {
  children: React.ReactNode;
  onClick?: any;
  component: any | "button" | "a";
  to?: string;
  disabled?: boolean;
  tabIndex?: string | number;
  href?: string;
  type?: string;
  withNewTab?: boolean;
  className?: string;
  style?: object;
  startAdornment?: React.ReactNode;
  endAdornment?: React.ReactNode;
};

export const classes = {
  root: "ButtonBase"
};

const ButtonBaseStyled = styled.button`
  color: inherit;
  cursor: pointer;
  margin: 0;
  border: 0;
  display: inline-flex;
  position: relative;
  user-select: none;
  align-items: center;
  border-radius: 0;
  vertical-align: middle;
  justify-content: center;
  text-decoration: none;
  background-color: transparent;
  font: inherit;
  padding: 0;
  outline: none;

  &:disabled,
  &:disabled:hover,
  &:disabled:active,
  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    cursor: default;
    pointer-events: none;
  }
`;

const ButtonBase = React.forwardRef<BaseButtonProps, any>(function ButtonBase(
  props,
  ref
) {
  const {
    children,
    component: componentProp = "button",
    tabIndex = "0",
    className,
    disabled,
    style,
    onClick,
    startAdornment,
    endAdornment,
    ...other
  } = props;

  let Component = componentProp;

  if (Component === "button" && other.href) {
    Component = "a";
  }

  if (Component === "button" && other.to) {
    Component = Link;
  }

  const buttonProps: any = {};

  // Sometimes it's helpful to use dataset props. Thus we need to make sure we apply them to the button
  Object.keys(other).forEach(key => {
    if (key.startsWith("data-")) {
      buttonProps[key] = other[key];
    }
  });

  if (Component === "button") {
    buttonProps.type = other.type;
    buttonProps.disabled = disabled;
  } else {
    if (Component !== "a" || !other.href) {
      buttonProps.role = "button";
    }
    buttonProps["aria-disabled"] = disabled;
  }

  function handleClick(ev: SyntheticEvent<any>) {
    ev.currentTarget.blur();

    if (typeof onClick === "function") {
      onClick(ev);
    }
  }

  return (
    <ButtonBaseStyled
      onClick={handleClick}
      as={Component}
      ref={ref}
      tabIndex={disabled ? "-1" : tabIndex}
      target={other && other.withNewTab ? "_blank" : null}
      style={style}
      className={clsx(className, classes.root)}
      {...buttonProps}
      {...other}
    >
      {startAdornment}
      {children}
      {endAdornment}
    </ButtonBaseStyled>
  );
});

export default ButtonBase;
