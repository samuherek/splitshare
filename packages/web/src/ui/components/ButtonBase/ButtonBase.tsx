import { Link } from '@reach/router';
import clsx from 'clsx';
import React, { SyntheticEvent } from 'react';
import styled from 'styled-components';
import { BaseProps } from '../../core/types';

export interface ButtonBaseProps extends BaseProps {
  onClick?: any;
  component?: React.ReactNode;
  to?: string;
  disabled?: boolean;
  tabIndex?: string | number;
  href?: string;
  type?: string;
  startAdornment?: React.ReactNode;
  endAdornment?: React.ReactNode;
}

export const classes = {
  root: 'ButtonBase',
};

const ButtonBaseStyled = styled.button`
  color: inherit;
  cursor: pointer;
  margin: 0;
  border: 0;
  display: inline-flex;
  position: relative;
  user-select: none;
  align-items: center;
  border-radius: 0;
  vertical-align: middle;
  justify-content: center;
  text-decoration: none;
  background-color: transparent;
  font: inherit;
  padding: 0;
  outline: none;

  &:disabled,
  &:disabled:hover,
  &:disabled:active,
  &.disabled,
  &.disabled:hover,
  &.disabled:active {
    cursor: default;
    pointer-events: none;
  }
`;

const ButtonBase = React.forwardRef<any, ButtonBaseProps>(function ButtonBase(
  props,
  ref
) {
  const {
    children,
    component: componentProp = 'button',
    tabIndex = '0',
    className,
    disabled,
    onClick,
    startAdornment,
    endAdornment,
    ...other
  } = props;

  let Component = componentProp;

  if (Component === 'button' && other.href) {
    Component = 'a';
  }

  if (Component === 'button' && other.to) {
    Component = Link;
  }

  const buttonProps: any = {};

  // Sometimes it's helpful to use dataset props. Thus we need to make sure we apply them to the button
  Object.keys(other).forEach(key => {
    if (key.startsWith('data-')) {
      // @ts-ignore
      buttonProps[key] = other[key];
    }
  });

  if (Component === 'button') {
    buttonProps.type = other.type;
    buttonProps.disabled = disabled;
  } else {
    if (Component !== 'a' || !other.href) {
      buttonProps.role = 'button';
    }
    buttonProps['aria-disabled'] = disabled;
  }

  function handleClick(ev: SyntheticEvent<any>) {
    ev.currentTarget.blur();

    if (typeof onClick === 'function') {
      onClick(ev);
    }
  }

  return (
    <ButtonBaseStyled
      onClick={handleClick}
      as={Component}
      ref={ref}
      tabIndex={disabled ? '-1' : tabIndex}
      // @ts-ignore
      target={other && other.withNewTab ? '_blank' : null}
      className={clsx(className, classes.root)}
      {...buttonProps}
      {...other}
    >
      {startAdornment}
      {children}
      {endAdornment}
    </ButtonBaseStyled>
  );
});

export default ButtonBase;
