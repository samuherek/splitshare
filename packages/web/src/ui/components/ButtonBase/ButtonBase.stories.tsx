import React from 'react';
import { boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import ButtonBase from './ButtonBase';

export default {
  title: 'Buttons/Button',
  component: ButtonBase,
};

export const Base = () => (
  <ButtonBase
    disabled={boolean('disabled', false)}
    onClick={action('Base button clicked')}
  >
    Button Base
  </ButtonBase>
);
