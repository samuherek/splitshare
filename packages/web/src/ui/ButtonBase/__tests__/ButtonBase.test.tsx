import { render } from '../../../test-utils/test-utils';
import React from 'react';
// import ButtonAdornment from "../../ButtonAdornment";
import ButtonBase from '../ButtonBase';

test('should match snapshots', () => {
  const emptyButton = render(<ButtonBase />);
  expect(emptyButton.container.firstChild).toMatchSnapshot();

  const textButton = render(<ButtonBase>Button</ButtonBase>);
  expect(textButton.container.firstChild).toMatchSnapshot();
});

test('should apply default styles', () => {
  const { container } = render(<ButtonBase />);
  const button = container.firstChild;
  // @ts-ignore
  expect(button).toHaveStyleRule('color', 'inherit');
  // @ts-ignore
  expect(button).toHaveStyleRule('cursor', 'pointer');
  // @ts-ignore
  expect(button).toHaveStyleRule('margin', '0');
  // @ts-ignore
  expect(button).toHaveStyleRule('padding', '0');
  // @ts-ignore
  expect(button).toHaveStyleRule('border', '0');
  // @ts-ignore
  expect(button).toHaveStyleRule('font', 'inherit');
  // @ts-ignore
  expect(button).toHaveStyleRule('font', 'inherit');
  // @ts-ignore
  expect(button).toHaveStyleRule('cursor', 'pointer');
  // @ts-ignore
  expect(button).not.toHaveStyleRule('opacity');
});

test('should change the button type', () => {
  const { getByText } = render(<ButtonBase type="submit">Button</ButtonBase>);
  expect(getByText(/button/i)).toHaveProperty('type', 'submit');
});

test('should change the button component and add accessibility requirements', () => {
  const { getByRole } = render(
    <ButtonBase component="span" role="checkbox" aria-checked={false} />
  );
  const checkbox = getByRole('checkbox');
  expect(checkbox).toHaveProperty('nodeName', 'SPAN');
  expect(checkbox).toHaveProperty('tabIndex', 0);
});

test('should not apply role="button" if type="button"', () => {
  const { getByText } = render(<ButtonBase type="button">Button</ButtonBase>);
  expect(getByText(/button/i)).not.toHaveAttribute('role');
});

test('should automatically change the button to an anochr element when href is provided', () => {
  const href = 'https://google.com';
  const { getByText } = render(<ButtonBase href={href}>Button</ButtonBase>);
  const button = getByText(/button/i);

  expect(button).toHaveProperty('nodeName', 'A');
  expect(button).not.toHaveProperty('role');
  expect(button).toHaveAttribute('href', href);
});

test('applies role="button" when an anchor is used without href', () => {
  const { getByRole } = render(<ButtonBase component="a">Button</ButtonBase>);
  const button = getByRole('button');

  expect(button).toHaveProperty('nodeName', 'A');
  expect(button).not.toHaveAttribute('type');
});

test('should not use an anchor element if explicit component and href is passed', () => {
  const href = 'https://google.com';
  const { getByRole } = render(
    <ButtonBase component="span" href={href}>
      Button
    </ButtonBase>
  );
  const button = getByRole('button');

  expect(button).toHaveProperty('nodeName', 'SPAN');
  expect(button).toHaveAttribute('href', href);
});

// test('should use react router <Link /> component if provided "to"', () => {
//   const path = '/';
//   const { getByText } = render(
//     withRouter(<ButtonBase to={path}>Button</ButtonBase>)
//   );
//   const button = getByText(/button/i);

//   expect(button).toHaveProperty('nodeName', 'A');
//   expect(button).toHaveAttribute('href', path);
// });

test('prop: disabled -> should have disabled styles', () => {
  const { getByText } = render(<ButtonBase disabled>Button</ButtonBase>);
  const button = getByText(/button/i);

  // @ts-ignore
  expect(button).toHaveStyleRule('cursor', 'default', {
    modifier: ':disabled',
  });
  // @ts-ignore
  expect(button).toHaveStyleRule('pointer-events', 'none', {
    modifier: ':disabled',
  });
});

test('prop: disabled -> should have a negative tabIndex', () => {
  const { getByText } = render(<ButtonBase disabled>Button</ButtonBase>);
  expect(getByText(/button/i)).toHaveProperty('tabIndex', -1);
});

test('prop: disabled -> should forward it to native buttons', () => {
  const { getByText } = render(
    <ButtonBase disabled component="button">
      Button
    </ButtonBase>
  );
  expect(getByText(/button/i)).toHaveProperty('disabled', true);
});

test('prop: disabled -> should use aria attributes for other components', () => {
  const { getByRole } = render(
    <ButtonBase component="span" disabled>
      Button
    </ButtonBase>
  );
  const button = getByRole('button');
  expect(button).not.toHaveAttribute('disabled');
  expect(button).toHaveAttribute('aria-disabled', 'true');
});

// test("prop: startAdornment and endAdornment should match snapshots", () => {
//   const start = render(
//     <ButtonBase
//       placeholder="Start"
//       startAdornment={<ButtonAdornment position="start">Kg</ButtonAdornment>}
//     />
//   );

//   expect(start.container.firstChild).toMatchSnapshot();

//   const end = render(
//     <ButtonBase
//       placeholder="End"
//       endAdornment={<ButtonAdornment position="end">Kg</ButtonAdornment>}
//     />
//   );

//   expect(end.container.firstChild).toMatchSnapshot();

//   const startEnd = render(
//     <ButtonBase
//       placeholder="Start and End"
//       startAdornment={<ButtonAdornment position="start">Kg</ButtonAdornment>}
//       endAdornment={<ButtonAdornment position="end">Kg</ButtonAdornment>}
//     />
//   );

//   expect(startEnd.container.firstChild).toMatchSnapshot();
// });

// test("prop: startAdornment should render adornment", () => {
//   const { getByText } = render(
//     <ButtonBase
//       placeholder="Start"
//       startAdornment={<ButtonAdornment position="start">Kg</ButtonAdornment>}
//     />
//   );
//   expect(getByText(/kg/i)).toBeDefined();
// });

// test("prop: endAdornment should render adornment", () => {
//   const { getByText } = render(
//     <ButtonBase
//       placeholder="End"
//       endAdornment={<ButtonAdornment position="end">Kg</ButtonAdornment>}
//     />
//   );
//   expect(getByText(/kg/i)).toBeDefined();
// });
