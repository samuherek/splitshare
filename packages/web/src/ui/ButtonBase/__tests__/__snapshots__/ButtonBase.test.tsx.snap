// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshots 1`] = `
.c0 {
  color: inherit;
  cursor: pointer;
  margin: 0;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0;
  vertical-align: middle;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  font: inherit;
  padding: 0;
  outline: none;
}

.c0:disabled,
.c0:disabled:hover,
.c0:disabled:active,
.c0.disabled,
.c0.disabled:hover,
.c0.disabled:active {
  cursor: default;
  pointer-events: none;
}

<button
  class="c0 ButtonBase"
  tabindex="0"
/>
`;

exports[`should match snapshots 2`] = `
.c0 {
  color: inherit;
  cursor: pointer;
  margin: 0;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0;
  vertical-align: middle;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  font: inherit;
  padding: 0;
  outline: none;
}

.c0:disabled,
.c0:disabled:hover,
.c0:disabled:active,
.c0.disabled,
.c0.disabled:hover,
.c0.disabled:active {
  cursor: default;
  pointer-events: none;
}

<button
  class="c0 ButtonBase"
  tabindex="0"
>
  Button
</button>
`;
